src/main/scala/is/hail/expr/ir/Compile.scala

```diff
@@ -0,0 +1,346 @@
+            i := 0))
+          val (mzero, vzero) = expression(zero)
+          fb.emit(Code(
+            xmout := mzero,
+            xvout := coerce[t](xmout.mux(dummyValue(typ), vzero)),
+            ltop))
+          (i < len).toConditional.emitConditional(fb.l, lnext, lend)
+          fb.emit(Code(
+            lnext,
+            xmv := !tarray.isElementDefined(region, xa, i),
+            xvv := coerce[u](xmv.mux(
+              dummyValue(tarray.elementType),
+              region.loadAnnotation(tarray.elementType)(tarray.loadElement(region, xa, i))))))
+          val (mbody, vbody) = expression(body, env = bodyenv)
+          fb.emit(Code(
+            xmout := mbody,
+            xvout := coerce[t](xmout.mux(dummyValue(typ), vbody)),
+            i := i + 1,
+            new JumpInsnNode(GOTO, ltop),
+            lend,
+            lmissing))
+          (xmout, xvout)
+        }
+      case ArrayFold(_,  _, _, _) =>
+        throw new UnsupportedOperationException(s"bad arrayfold $ir")
+      case MakeStruct(fields, missingness) =>
+        assert(missingness != null, "run explicit missingness first")
+        val t = TStruct(fields.map { case (name, t, _) => (name, t) }: _*)
+        val initializers = fields.map { case (_, t, v) => (t, expression(v)) }
+        val srvb = new StagedRegionValueBuilder(fb, t)
+
+        present(Code(
+          srvb.start(false),
+          Code(initializers.map { case (t, (mv, vv)) =>
+            Code(
+              mv.mux(srvb.setMissing(), srvb.addAnnotation(t)(vv)),
+              srvb.advance()) }: _*),
+          srvb.offset))
+      case GetField(o, name, _) =>
+        val t = o.typ.asInstanceOf[TStruct]
+        val fieldIdx = t.fieldIdx(name)
+        val (mo, vo) = expression(o)
+        val xmo = mb.newBit()
+        val xo = fb.newLocal[Long]
+        fb.emit(xmo := mo)
+        fb.emit(xo := coerce[Long](xmo.mux(dummyValue(t), vo)))
+        (xmo || !t.isFieldDefined(region, xo, fieldIdx),
+          region.loadAnnotation(t)(t.fieldOffset(xo, fieldIdx)))
+      case GetFieldMissingness(o, name) =>
+        val t = o.typ.asInstanceOf[TStruct]
+        val fieldIdx = t.fieldIdx(name)
+        val (mo, vo) = expression(o)
+        val xmo = mb.newBit()
+        val xo = fb.newLocal[Long]
+        fb.emit(xmo := mo)
+        fb.emit(xo := coerce[Long](xmo.mux(dummyValue(t), vo)))
+        (xmo, !t.isFieldDefined(region, xo, fieldIdx))
+      case Seq(stmts, typ) =>
+        ???
+      case In(i, typ) =>
+        (fb.getArg[Boolean](i*2 + 3), fb.getArg(i*2 + 2)(typeToTypeInfo(typ)))
+      case InMissingness(i) =>
+        present(fb.getArg[Boolean](i*2 + 3))
+      case Out(v) =>
+        val (mv, vv) = expression(v)
+        typeToTypeInfo(v.typ) match { case ti: TypeInfo[t] =>
+          present(mv.mux(
+            Code._throw(Code.newInstance[RuntimeException, String]("cannot return empty")),
+            Code._return(coerce[t](vv))(ti)))
```

Comment:
See my comment about emit vs return.  I feel like this should be emitted and not returned.